### PR TITLE
Use WKWebView also for redirects in the payment process.

### PR DESCRIPTION
**Pull request by IhmeHippi**

### CODE REVIEW

File: Podfile [hunk 1]
```diff
@@ -6,6 +6,8 @@ target 'SwedbankPaySDK' do
   use_frameworks!
 
   # Pods for SwedbankPaySDK
+  pod 'Alamofire', '~> 4.9.0'
+  pod 'AlamofireObjectMapper', '~> 5.0'
```
Comment (IhmeHippi) [inline Podfile]: Without these the library project fails to build on its own. This is needed for unit tests.

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 2]
```diff
@@ -13,94 +13,227 @@
...
+            appendLiteral(SwedbankWebView.emitCallback(event: event.rawValue, argument: argument))
+        }
+        
+        fileprivate mutating func appendInterpolation(_ component: SwedbankWebView.TemplateComponent) {
+            components.append(component)
+        }
     }
 }
 
+private extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
private extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 3]
```diff
@@ -13,94 +13,227 @@
...
-                    },
-                    onError: function(error) {
-                        webkit.messageHandlers.onPaymentError.postMessage(event);
-                    }
-                }).open();
-            </script>
-            </body>
-            </html>
-        """
+extension SwedbankWebView.HTMLTemplate : ExpressibleByStringInterpolation {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
extension SwedbankPayWebView.HTMLTemplate : ExpressibleByStringInterpolation {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 4]
```diff
@@ -13,94 +13,227 @@
...
-        """
+        </head>
+        <body>
+            <div id="checkout" />
+        </body>
+    </html>
+    """
+}
+
+private extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
private extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 5]
```diff
@@ -13,94 +13,227 @@
...
+            </script>
+        </head>
+        <body>
+            <div id="checkin" />
+        </body>
+    </html>
+    """
+}
 
+extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 6]
```diff
@@ -13,94 +13,227 @@
...
+            return strings.joined()
+        }
+        
+        func createScriptMessageHandler(eventHandler: @escaping (T, Any?) -> Void) -> WKScriptMessageHandler {
+            return CallbackScriptMessageHandler(eventHandler: eventHandler)
+        }
+    }
+}
+
+extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 7]
```diff
@@ -13,94 +13,227 @@
...
+
+private let paymentMenuDelay = "500"
+
+enum SwedbankWebView {}
+
+extension SwedbankWebView {
+    static let scriptMessageHandlerName = "swedbankpay"
+}
+
+extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 8]
```diff
@@ -13,94 +13,227 @@
...
 // limitations under the License.
 
-class SwedbankWebView {
+import WebKit
+
+private let paymentMenuDelay = "500"
+
+enum SwedbankWebView {}
+
+extension SwedbankWebView {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
extension SwedbankPayWebView {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 9]
```diff
@@ -13,94 +13,227 @@
 // See the License for the specific language governing permissions and
 // limitations under the License.
 
-class SwedbankWebView {
+import WebKit
+
+private let paymentMenuDelay = "500"
+
+enum SwedbankWebView {}
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
enum SwedbankPayWebView {}
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 10]
```diff
@@ -143,56 +233,32 @@ public final class SwedbankPaySDKController: UIViewController {
         let operationType = Operation.TypeString.viewPaymentOrder.rawValue
         if let jsURL: String = list.operations.first(where: {$0.contentType == "application/javascript" && $0.rel == operationType})?.href {
             viewModel.viewPaymentOrderLink = jsURL
-            loadWebViewURL(jsURL, type: .paymentOrder)
+            loadPage(template: SwedbankWebView.paymentTemplate, scriptUrl: jsURL) { [weak self] (event, argument) in
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
            loadPage(template: SwedbankPayWebView.paymentTemplate, scriptUrl: jsURL) { [weak self] (event, argument) in
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 11]
```diff
@@ -107,19 +184,30 @@ public final class SwedbankPaySDKController: UIViewController {
...
+    private func reloadPaymentMenuIfAtRoot() {
+        if rootWebViewController.isAtRoot == true {
+            reloadPaymentMenu()
         }
     }
     
+    private func reloadPaymentMenu(delay: Bool = false) {
+        if let link = viewModel.viewPaymentOrderLink {
+            dismissExtraWebViews(reloadPaymentMenuIfAtRoot: false)
+            loadPage(template: SwedbankWebView.paymentTemplate, scriptUrl: link, delay: delay) { [weak self] (event, argument) in
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
            loadPage(template: SwedbankPayWebView.paymentTemplate, scriptUrl: link, delay: delay) { [weak self] (event, argument) in
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 12]
```diff
@@ -130,7 +218,9 @@ public final class SwedbankPaySDKController: UIViewController {
     private func createConsumerURL(_ list: OperationsList) {
         let operationType = Operation.TypeString.viewConsumerIdentification.rawValue
         if let jsURL: String = list.operations.first(where: {$0.contentType == "application/javascript" && $0.rel == operationType})?.href {
-            loadWebViewURL(jsURL, type: .consumerIdentification)
+            loadPage(template: SwedbankWebView.checkInTemplate, scriptUrl: jsURL) { [weak self] (event, argument) in
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
            loadPage(template: SwedbankPayWebView.checkInTemplate, scriptUrl: jsURL) { [weak self] (event, argument) in
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 13]
```diff
@@ -143,56 +233,32 @@ public final class SwedbankPaySDKController: UIViewController {
...
-            html = SwedbankWebView.createCheckoutHTML(url)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onPaymentMenuInstrumentSelected.rawValue)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onPaymentCompleted.rawValue)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onPaymentFailed.rawValue)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onPaymentCreated.rawValue)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onPaymentToS.rawValue)
-            contentController.add(messageHandler, name: SwedbankWebView.PaymentEvent.onError.rawValue)
+    private func set(scriptMessageHandler: WKScriptMessageHandler?) {
+        debugPrint("setting script message handler to \(scriptMessageHandler == nil ? "" : "non-")nil")
+        userContentController.removeScriptMessageHandler(forName: SwedbankWebView.scriptMessageHandlerName)
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        userContentController.removeScriptMessageHandler(forName: SwedbankPayWebView.scriptMessageHandlerName)
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 14]
```diff
@@ -143,56 +233,32 @@ public final class SwedbankPaySDKController: UIViewController {
...
+        if let scriptMessageHandler = scriptMessageHandler {
+            userContentController.add(scriptMessageHandler, name: SwedbankWebView.scriptMessageHandlerName)
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
            userContentController.add(scriptMessageHandler, name: SwedbankPayWebView.scriptMessageHandlerName)
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 15]
```diff
@@ -143,56 +233,32 @@ public final class SwedbankPaySDKController: UIViewController {
...
         }
+    }
+    
+    private func loadPage<T>(template: SwedbankWebView.HTMLTemplate<T>, scriptUrl: String, delay: Bool = false, eventHandler: @escaping (T, Any?) -> Void) {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
    private func loadPage<T>(template: SwedbankPayWebView.HTMLTemplate<T>, scriptUrl: String, delay: Bool = false, eventHandler: @escaping (T, Any?) -> Void) {
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 16]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
 
 /// Extension to handle the WKWebview JavaScript events
 private extension SwedbankPaySDKController {
-    private class MyScriptMessageHandler: NSObject, WKScriptMessageHandler {
-        private weak var owner: SwedbankPaySDKController?
-        
-        init(owner: SwedbankPaySDKController) {
-            super.init()
-            self.owner = owner
+    func on(consumerEvent: SwedbankWebView.ConsumerEvent, argument: Any?) {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
    func on(consumerEvent: SwedbankPayWebView.ConsumerEvent, argument: Any?) {
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 17]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
-                handleToSEvent(message.body)
-            case SwedbankWebView.PaymentEvent.onError.rawValue:
-                let msg: String = message.body as? String ?? "Unknown error"
-                owner?.paymentFailed(SwedbankPaySDK.Problem.Client(.MobileSDK(.InvalidRequest(message: msg, raw: nil))))
-            default:
-                debugPrint("SwedbankPaySDK: undefined event received")
-            }
+    }
+    
+    func on(paymentEvent: SwedbankWebView.PaymentEvent, argument: Any?) {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
    func on(paymentEvent: SwedbankPayWebView.PaymentEvent, argument: Any?) {
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 18]
```diff
@@ -13,94 +13,227 @@
...
+extension SwedbankWebView.HTMLTemplate : ExpressibleByStringInterpolation {
+    init(stringInterpolation: StringInterpolation) {
+        components = stringInterpolation.components
+    }
+    init(stringLiteral value: String) {
+        components = [.literal(value)]
+    }
+    
+    struct StringInterpolation : StringInterpolationProtocol {
+        fileprivate var components: [SwedbankWebView.TemplateComponent] = []
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
        fileprivate var components: [SwedbankPayWebView.TemplateComponent] = []
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 19]
```diff
@@ -13,94 +13,227 @@
...
+        init(literalCapacity: Int, interpolationCount: Int) {
+            components.reserveCapacity(2 * interpolationCount + 1)
+        }
+        
+        mutating func appendLiteral(_ literal: String) {
+            components.append(.literal(literal))
+        }
+        
+        mutating func appendInterpolation(_ event: T, _ argument: String) {
+            appendLiteral(SwedbankWebView.emitCallback(event: event.rawValue, argument: argument))
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
            appendLiteral(SwedbankPayWebView.emitCallback(event: event.rawValue, argument: argument))
```

File: SwedbankPaySDK/Classes/Models/SwedbankWebView.swift [hunk 20]
```diff
@@ -13,94 +13,227 @@
...
+        }
+        
+        fileprivate mutating func appendInterpolation(_ component: SwedbankWebView.TemplateComponent) {
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/Models/SwedbankWebView.swift]:
```suggestion
        fileprivate mutating func appendInterpolation(_ component: SwedbankPayWebView.TemplateComponent) {
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 21]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+        switch paymentEvent {
+        case .onScriptLoaded:
+            initialLoadingIndicator.stopAnimating()
+        case .onScriptError:
+            debugPrint("Script \(String(describing: argument)) failed to load")
+            paymentFailed(.Server(.UnexpectedContent(status: -1, contentType: nil, body: nil))) // TODO: Better error
+        case SwedbankWebView.PaymentEvent.onPaymentMenuInstrumentSelected:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onPaymentMenuInstrumentSelected:
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 22]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+            debugPrint("SwedbankPaySDK: onPaymentMenuInstrumentSelected event received")
+        case SwedbankWebView.PaymentEvent.onPaymentCompleted:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onPaymentCompleted:
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 23]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+            debugPrint("SwedbankPaySDK: onPaymentCompleted event received")
+            paymentComplete()
+        case SwedbankWebView.PaymentEvent.onPaymentFailed:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onPaymentFailed:
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 24]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+            let msg: String = argument as? String ?? "Unknown error"
+            paymentFailed(SwedbankPaySDK.Problem.Client(.MobileSDK(.InvalidRequest(message: msg, raw: nil))))
+        case SwedbankWebView.PaymentEvent.onPaymentCreated:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onPaymentCreated:
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 25]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+            debugPrint("SwedbankPaySDK: onPaymentCreated event received")
+        case SwedbankWebView.PaymentEvent.onPaymentToS:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onPaymentToS:
```

File: SwedbankPaySDK/Classes/SwedbankPaySDKController.swift [hunk 26]
```diff
@@ -271,78 +287,77 @@ extension SwedbankPaySDKController: WKNavigationDelegate {
...
+            handleToSEvent(argument)
+        case SwedbankWebView.PaymentEvent.onError:
```
Comment (asbjornu) [inline SwedbankPaySDK/Classes/SwedbankPaySDKController.swift]:
```suggestion
        case SwedbankPayWebView.PaymentEvent.onError:
```

**Pull request merged**